doc/man-pages/man1/cmdebug.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "CMDEBUG 1"
.TH CMDEBUG 1 "2010-02-11" "OpenAFS" "AFS Command Reference"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
cmdebug \- Reports the status of a host Cache Manager
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBcmdebug\fR \fB\-servers\fR\ <\fImachine\fR> [\fB\-port\fR\ <\fI\s-1IP\s0\ port\fR>]
    [\fB\-long\fR] [\fB\-refcounts\fR] [\fB\-callbacks\fR] [\fB\-ctime\fR] [\fB\-addrs\fR]
    [\fB\-cache\fR] [\fB\-cellservdb\fR] [\fB\-help\fR]
.PP
\&\fBcmdebug\fR \fB\-s\fR\ <\fImachine\fR> [\fB\-p\fR\ <\fI\s-1IP\s0\ port\fR>] [\fB\-l\fR] [\fB\-r\fR] [\fB\-cal\fR]
    [\fB\-ct\fR] [\fB\-a\fR] [\fB\-cac\fR] \fB\-ce\fR] [\fB\-h\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBcmdebug\fR command displays information about the Cache Manager
and client cache status on a local or remote \s-1AFS\s0 client machine.  By
default, it displays all locked cache entries, but other information
can be requested via various options.
.SH "CAUTIONS"
.IX Header "CAUTIONS"
The \fB\-ctime\fR option is only available with OpenAFS version 1.4.7 and
later or version 1.5.28 or later.  This option can be used to gather
information from any version of the Unix OpenAFS client, but can only
query Windows clients running OpenAFS version 1.5.28 or later.
.PP
The \fB\-cellservdb\fR option is only available with OpenAFS version 1.4.7
and later or version 1.5.31 or later.  This option can be used to gather
information from any version of the Unix OpenAFS client, but can only
query Windows clients running OpenAFS version 1.5.31 or later.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-servers\fR <\fImachine\fR>" 4
.IX Item "-servers <machine>"
Names the client machine for which to display Cache Manager status.
Provide the machine's \s-1IP\s0 address in dotted decimal format, its fully
qualified host name (for example, \fBfs1.example.com\fR), or the shortest
abbreviated form of its host name that distinguishes it from other
machines. Successful use of an abbreviated form depends on the
availability of a name resolution service (such as the Domain Name
Service or a local host table) at the time the command is issued.
.IP "\fB\-port\fR <\fI\s-1IP\s0 port\fR>" 4
.IX Item "-port <IP port>"
Identifies the port on which to contact the Cache Manager.  By default,
the standard port 7001 is used, so this option is very rarely needed.
.IP "\fB\-long\fR" 4
.IX Item "-long"
Reports on all lock statuses and all cache entries, rather than only
locked cache entries.  Do not use this option with \fB\-refcounts\fR,
\&\fB\-callbacks\fR, \fB\-addrs\fR, \fB\-cache\fR, or \fB\-cellservdb\fR.
.IP "\fB\-refcounts\fR" 4
.IX Item "-refcounts"
Reports only those cache entries with non-zero reference counts.  Do not
use this option with \fB\-long\fR, \fB\-callbacks\fR, \fB\-addrs\fR, or \fB\-cache\fR.
.IP "\fB\-callbacks\fR" 4
.IX Item "-callbacks"
Reports only those cache entries with callbacks.  Do not use this option
with \fB\-long\fR, \fB\-refcounts\fR, \fB\-addrs\fR, or \fB\-cache\fR.
.IP "\fB\-ctime\fR" 4
.IX Item "-ctime"
Causes entry expiration times to be shown in human-readable format. Do
not use this option with \fB\-addrs\fR or \fB\-cache\fR.
.IP "\fB\-addrs\fR" 4
.IX Item "-addrs"
Rather than showing any cache entries, displays the interfaces the Cache
Manager answers on, including their netmasks and MTUs.  This is useful for
analyzing clients that are multihomed and identifying problems with
netmasks or \s-1MTU\s0 settings.  Do not use this option with \fB\-long\fR,
\&\fB\-refcounts\fR, \fB\-callbacks\fR, or \fB\-cache\fR.
.IP "\fB\-cache\fR" 4
.IX Item "-cache"
Rather than showing any cache entries, displays the cache configuration
for the client machine.  The information displayed is essentially the
information that can be configured via parameters to \fBafsd\fR.  Do not use
this option with \fB\-long\fR, \fB\-refcounts\fR, \fB\-callbacks\fR, or \fB\-addrs\fR.
.IP "\fB\-cellservdb\fR" 4
.IX Item "-cellservdb"
Lists all known volume location database records in a
CellServDB-compatible format. This includes all records in memory,
including those from the CellServDB file, \s-1AFSDB\s0 \s-1DNS\s0 records, and the
\&\fBfs newcell\fR command. This option could be used to see if a client
has the latest copy of the CellServDB file. Do not use this option
with \fB\-long\fR, \fB\-refcounts\fR, \fB\-callbacks\fR, or \fB\-cache\fR.
.IP "\fB\-help\fR" 4
.IX Item "-help"
Prints the online help for this command. All other valid options are
ignored.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Displays all of the locked cache entries on \f(CW\*(C`client1\*(C'\fR:
.PP
.Vb 1
\&    % cmdebug client1
.Ve
.PP
Displays the cache configuration for \f(CW\*(C`client1.example.com\*(C'\fR:
.PP
.Vb 1
\&    % cmdebug client1.example.com \-cache
.Ve
.PP
Displays all cache entries for \f(CW\*(C`client2.example.com\*(C'\fR:
.PP
.Vb 1
\&    % cmdebug client2.example.com \-long
.Ve
.SH "PRIVILEGE REQUIRED"
.IX Header "PRIVILEGE REQUIRED"
None
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIafsd\fR\|(8),
\&\fICellServDB\fR\|(5),

NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.


src/venus/cmdebug.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

#include <afsconfig.h>
#include <afs/param.h>



#include <sys/types.h>
#ifdef AFS_NT40_ENV
#include <winsock2.h>
#include <rpc.h>
#else
#ifdef HAVE_NETINET_IN_H
#include <netinet/in.h>
#endif
#include <sys/socket.h>
#include <netdb.h>
#endif
#include <stdio.h>
#include <string.h>
#ifdef	AFS_AIX32_ENV
#include <signal.h>
#endif
#include <afs/afscbint.h>
#include <afs/cmd.h>
#include <rx/rx.h>
#include <lock.h>
#include <afs/afs_args.h>
#include <afs/afsutil.h>
#include <afs/com_err.h>

static int print_ctime = 0;

static int
PrintCacheConfig(struct rx_connection *aconn)
{
    struct cacheConfig c;
    afs_uint32 srv_ver, conflen;
    int code;

    c.cacheConfig_len = 0;
    c.cacheConfig_val = NULL;
    code = RXAFSCB_GetCacheConfig(aconn, 1, &srv_ver, &conflen, &c);
    if (code) {
	printf("cmdebug: error checking cache config: %s\n",
	       afs_error_message(code));
	return 0;
    }

    if (srv_ver == AFS_CLIENT_RETRIEVAL_FIRST_EDITION) {
	struct cm_initparams_v1 *c1;

	if (c.cacheConfig_len != sizeof(*c1) / sizeof(afs_uint32)) {
	    printf("cmdebug: configuration data size mismatch (%d != %" AFS_SIZET_FMT ")\n",
		   c.cacheConfig_len, sizeof(*c1) / sizeof(afs_uint32));
	    return 0;
	}

	c1 = (struct cm_initparams_v1 *)c.cacheConfig_val;
	printf("Chunk files:   %d\n", c1->nChunkFiles);
	printf("Stat caches:   %d\n", c1->nStatCaches);
	printf("Data caches:   %d\n", c1->nDataCaches);
	printf("Volume caches: %d\n", c1->nVolumeCaches);
	printf("Chunk size:    %d", c1->otherChunkSize);
	if (c1->firstChunkSize != c1->otherChunkSize)
	    printf(" (first: %d)", c1->firstChunkSize);
	printf("\n");
	printf("Cache size:    %d kB\n", c1->cacheSize);
	printf("Set time:      %s\n", c1->setTime ? "yes" : "no");
	printf("Cache type:    %s\n", c1->memCache ? "memory" : "disk");
    } else {
	printf("cmdebug: unsupported server version %d\n", srv_ver);
    }
    return 0;
}

#ifndef CAPABILITY_BITS
#define CAPABILITY_ERRORTRANS (1<<0)
#define CAPABILITY_BITS 1
#endif

static int
PrintInterfaces(struct rx_connection *aconn)
{
    Capabilities caps;
    struct interfaceAddr addr;
#ifdef AFS_NT40_ENV
    char * p;
#else
    char uuidstr[128];
#endif
    int i, code;

    caps.Capabilities_val = NULL;
    caps.Capabilities_len = 0;

    code = RXAFSCB_TellMeAboutYourself(aconn, &addr, &caps);
    if (code == RXGEN_OPCODE)
        code = RXAFSCB_WhoAreYou(aconn, &addr);
    if (code) {
	printf("cmdebug: error checking interfaces: %s\n",
	       afs_error_message(code));
	return 0;
    }

#ifdef AFS_NT40_ENV
    UuidToString((UUID *)&addr.uuid, &p);
    printf("UUID: %s\n",p);
    RpcStringFree(&p);
#else
    afsUUID_to_string(&addr.uuid, uuidstr, sizeof(uuidstr));
    printf("UUID: %s\n",uuidstr);
#endif

    printf("Host interfaces:\n");
    for (i = 0; i < addr.numberOfInterfaces; i++) {
	printf("%s", afs_inet_ntoa(htonl(addr.addr_in[i])));
	if (addr.subnetmask[i])
	    printf(", netmask %s", afs_inet_ntoa(htonl(addr.subnetmask[i])));
	if (addr.mtu[i])
	    printf(", MTU %d", addr.mtu[i]);
	printf("\n");
    }

    if (caps.Capabilities_val) {
        printf("Capabilities:\n");
        if (caps.Capabilities_val[0] & CAPABILITY_ERRORTRANS) {
            printf("Error Translation\n");  
        }
        printf("\n");
    }

    if (caps.Capabilities_val)
	free(caps.Capabilities_val);
    caps.Capabilities_val = NULL;
    caps.Capabilities_len = 0;

    return 0;
}

static int
IsLocked(register struct AFSDBLockDesc *alock)
{
    if (alock->waitStates || alock->exclLocked || alock->numWaiting
	|| alock->readersReading)
	return 1;
    return 0;
}

static int
PrintLock(register struct AFSDBLockDesc *alock)
{
    printf("(");
    if (alock->waitStates) {
	if (alock->waitStates & READ_LOCK)
	    printf("reader_waiting");
	if (alock->waitStates & WRITE_LOCK)
	    printf("writer_waiting");
	if (alock->waitStates & SHARED_LOCK)
	    printf("upgrade_waiting");
    } else
	printf("none_waiting");
    if (alock->exclLocked) {
	if (alock->exclLocked & WRITE_LOCK)
	    printf(", write_locked");
	if (alock->exclLocked & SHARED_LOCK)
	    printf(", upgrade_locked");
	printf("(pid:%d at:%d)", alock->pid_writer, alock->src_indicator);
    }
    if (alock->readersReading)
	printf(", %d read_locks(pid:%d)", alock->readersReading,
	       alock->pid_last_reader);
    if (alock->numWaiting)
	printf(", %d waiters", alock->numWaiting);
    printf(")");
    return 0;
}

static int
PrintLocks(register struct rx_connection *aconn, int aint32)
{
    register int i;
    struct AFSDBLock lock;
    afs_int32 code;

    for (i = 0; i < 1000; i++) {
	code = RXAFSCB_GetLock(aconn, i, &lock);
	if (code) {
	    if (code == 1)
		break;
	    /* otherwise we have an unrecognized error */
	    printf("cmdebug: error checking locks: %s\n",
		   afs_error_message(code));
	    return code;
	}
	/* here we have the lock information, so display it, perhaps */
	if (aint32 || IsLocked(&lock.lock)) {
	    printf("Lock %s status: ", lock.name);
	    PrintLock(&lock.lock);
	    printf("\n");
	}
    }
    return 0;
}

struct cell_cache {
    afs_int32 cellnum;
    char *cellname;
    struct cell_cache *next;
};

static char *
GetCellName(struct rx_connection *aconn, afs_int32 cellnum)
{
    static int no_getcellbynum;
    static struct cell_cache *cache;
    struct cell_cache *tcp;
    int code;
    char *cellname;
    serverList sl;

    if (no_getcellbynum)
	return NULL;

    for (tcp = cache; tcp; tcp = tcp->next)
	if (tcp->cellnum == cellnum)
	    return tcp->cellname;

    cellname = NULL;
    sl.serverList_len = 0;
    sl.serverList_val = NULL;
    code = RXAFSCB_GetCellByNum(aconn, cellnum, &cellname, &sl);
    if (code) {
	if (code == RXGEN_OPCODE)
	    no_getcellbynum = 1;
	return NULL;
    }

    if (sl.serverList_val)
	free(sl.serverList_val);
    tcp = malloc(sizeof(struct cell_cache));
    tcp->next = cache;
    tcp->cellnum = cellnum;
    tcp->cellname = cellname;
    cache = tcp;

    return cellname;
}

static int
PrintCacheEntries32(struct rx_connection *aconn, int aint32)
{
    register int i;
    register afs_int32 code;
    struct AFSDBCacheEntry centry;
    char *cellname;

    for (i = 0; i < 1000000; i++) {
	code = RXAFSCB_GetCE(aconn, i, &centry);
	if (code) {
	    if (code == 1)
		break;
	    printf("cmdebug: failed to get cache entry %d (%s)\n", i,
		   afs_error_message(code));
	    return code;
	}

	if (centry.addr == 0) {
	    /* PS output */
	    printf("Proc %4d sleeping at %08x, pri %3d\n",
		   centry.netFid.Vnode, centry.netFid.Volume,
		   centry.netFid.Unique - 25);
	    continue;
	}

	if ((aint32 == 0 && !IsLocked(&centry.lock)) ||
            (aint32 == 2 && centry.refCount == 0) ||
            (aint32 == 4 && centry.callback == 0))
	    continue;

	/* otherwise print this entry */
	printf("** Cache entry @ 0x%08x for %d.%d.%d.%d", centry.addr,
	       centry.cell, centry.netFid.Volume, centry.netFid.Vnode,
	       centry.netFid.Unique);

	cellname = GetCellName(aconn, centry.cell);
	if (cellname)
	    printf(" [%s]\n", cellname);
	else
	    printf("\n");

	if (IsLocked(&centry.lock)) {
	    printf("    locks: ");
	    PrintLock(&centry.lock);
	    printf("\n");
	}
	printf("    %12d bytes  DV %12d  refcnt %5d\n", centry.Length,
	       centry.DataVersion, centry.refCount);
        if (print_ctime) {
            time_t t = centry.cbExpires;
            printf("    callback %08x\texpires %s\n", centry.callback,
                    ctime(&t));
        } else
            printf("    callback %08x\texpires %u\n", centry.callback,
                   centry.cbExpires);
	printf("    %d opens\t%d writers\n", centry.opens, centry.writers);

	/* now display states */
	printf("    ");
	if (centry.mvstat == 0)
	    printf("normal file");
	else if (centry.mvstat == 1)
	    printf("mount point");
	else if (centry.mvstat == 2)
	    printf("volume root");
	else if (centry.mvstat == 3)	/* windows */
	    printf("directory");
	else if (centry.mvstat == 4)	/* windows */
	    printf("symlink");
	else if (centry.mvstat == 5)	/* windows */
	    printf("microsoft dfs link");
	else if (centry.mvstat == 6)	/* windows */
	    printf("invalid link");
	else
	    printf("bogus mvstat %d", centry.mvstat);
	printf("\n    states (0x%x)", centry.states);
	if (centry.states & 1)
	    printf(", stat'd");
	if (centry.states & 2)
	    printf(", backup");
	if (centry.states & 4)
	    printf(", read-only");
	if (centry.states & 8)
	    printf(", mt pt valid");
	if (centry.states & 0x10)
	    printf(", pending core");
	if (centry.states & 0x40)
	    printf(", wait-for-store");
	if (centry.states & 0x80)
	    printf(", mapped");
	printf("\n");
    }
    return 0;
}

static int
PrintCacheEntries64(struct rx_connection *aconn, int aint32)
{
    register int i;
    register afs_int32 code;
    struct AFSDBCacheEntry64 centry;
    char *cellname;

    for (i = 0; i < 1000000; i++) {
	code = RXAFSCB_GetCE64(aconn, i, &centry);
	if (code) {
	    if (code == 1)
		break;
	    printf("cmdebug: failed to get cache entry %d (%s)\n", i,
		   afs_error_message(code));
	    return code;
	}

	if (centry.addr == 0) {
	    /* PS output */
	    printf("Proc %4d sleeping at %08x, pri %3d\n",
		   centry.netFid.Vnode, centry.netFid.Volume,
		   centry.netFid.Unique - 25);
	    continue;
	}

	if ((aint32 == 0 && !IsLocked(&centry.lock)) ||
            (aint32 == 2 && centry.refCount == 0) ||
            (aint32 == 4 && centry.callback == 0))
	    continue;

	/* otherwise print this entry */
	printf("** Cache entry @ 0x%08x for %d.%d.%d.%d", centry.addr,
	       centry.cell, centry.netFid.Volume, centry.netFid.Vnode,
	       centry.netFid.Unique);

	cellname = GetCellName(aconn, centry.cell);
	if (cellname)
	    printf(" [%s]\n", cellname);
	else
	    printf("\n");

	if (IsLocked(&centry.lock)) {
	    printf("    locks: ");
	    PrintLock(&centry.lock);
	    printf("\n");
	}
#ifdef AFS_64BIT_ENV
#ifdef AFS_NT40_ENV
	printf("    %12I64d bytes  DV %12d  refcnt %5d\n", centry.Length,
	       centry.DataVersion, centry.refCount);
#else
	printf("    %12llu bytes  DV %12d  refcnt %5d\n", centry.Length,
	       centry.DataVersion, centry.refCount);
#endif
#else
	printf("    %12d bytes  DV %12d  refcnt %5d\n", centry.Length,
	       centry.DataVersion, centry.refCount);
#endif
        if (print_ctime) {
            time_t t = centry.cbExpires;
            printf("    callback %08x\texpires %s\n", centry.callback,
                    ctime(&t));
        } else
            printf("    callback %08x\texpires %u\n", centry.callback,
                   centry.cbExpires);
	printf("    %d opens\t%d writers\n", centry.opens, centry.writers);

	/* now display states */
	printf("    ");
	if (centry.mvstat == 0)
	    printf("normal file");
	else if (centry.mvstat == 1)
	    printf("mount point");
	else if (centry.mvstat == 2)
	    printf("volume root");
	else if (centry.mvstat == 3)
	    printf("directory");
	else if (centry.mvstat == 4)
	    printf("symlink");
	else if (centry.mvstat == 5)
	    printf("microsoft dfs link");
	else if (centry.mvstat == 6)
	    printf("invalid link");
        else
	    printf("bogus mvstat %d", centry.mvstat);
	printf("\n    states (0x%x)", centry.states);
	if (centry.states & 1)
	    printf(", stat'd");
	if (centry.states & 2)
	    printf(", backup");
	if (centry.states & 4)
	    printf(", read-only");
	if (centry.states & 8)
	    printf(", mt pt valid");
	if (centry.states & 0x10)
	    printf(", pending core");
	if (centry.states & 0x40)
	    printf(", wait-for-store");
	if (centry.states & 0x80)
	    printf(", mapped");
	printf("\n");
    }
    return 0;
}

static int
PrintCacheEntries(struct rx_connection *aconn, int aint32)
{
    register afs_int32 code;
    struct AFSDBCacheEntry64 centry64;

    code = RXAFSCB_GetCE64(aconn, 0, &centry64);
    if (code != RXGEN_OPCODE)
	return PrintCacheEntries64(aconn, aint32);
    else
	return PrintCacheEntries32(aconn, aint32);
}

static int
PrintCellServDBEntry(struct rx_connection *aconn, afs_int32 cellnum)
{
    int code;
    char *cellname;
    serverList sl;
    unsigned int n;
    int rc = 0;

    cellname = NULL;
    sl.serverList_len = 0;
    sl.serverList_val = NULL;
    code = RXAFSCB_GetCellServDB(aconn, cellnum, &cellname, &sl);
    if (code)
	return 0;

    if ( !cellname || !cellname[0] )
        goto done;

    rc = 1;
    printf(">%-23s#%s\n", cellname, cellname);

    if (sl.serverList_val) {
        for ( n=0; n<sl.serverList_len; n++) {
            struct hostent *host;
            afs_uint32      addr = ntohl(sl.serverList_val[n]);

            host = gethostbyaddr((const char *)&addr, sizeof(afs_uint32), AF_INET);
            printf("%-28s#%s\n", afs_inet_ntoa(addr), 
                    host ? host->h_name : "");
        }
    }

  done:
    if (cellname)
        free(cellname);

    if (sl.serverList_val)
	free(sl.serverList_val);

    return rc;
}

static void
PrintCellServDB(struct rx_connection *aconn) 
{
    afs_int32 index;

    for ( index = 0 ; PrintCellServDBEntry(aconn, index); index++ );
}

int
CommandProc(struct cmd_syndesc *as, void *arock)
{
    struct rx_connection *conn;
    register char *hostName;
    register struct hostent *thp;
    afs_int32 port;
    struct rx_securityClass *secobj;
    int int32p;
    afs_int32 addr;

    hostName = as->parms[0].items->data;
    if (as->parms[1].items)
	port = atoi(as->parms[1].items->data);
    else
	port = 7001;
    thp = hostutil_GetHostByName(hostName);
    if (!thp) {
	printf("cmdebug: can't resolve address for host %s.\n", hostName);
	exit(1);
    }
    memcpy(&addr, thp->h_addr, sizeof(afs_int32));
    secobj = rxnull_NewServerSecurityObject();
    conn = rx_NewConnection(addr, htons(port), 1, secobj, 0);
    if (!conn) {
	printf("cmdebug: failed to create connection for host %s\n",
	       hostName);
	exit(1);
    }

    if (as->parms[6].items) {
	/* -addrs */
	PrintInterfaces(conn);
	return 0;
    }
    if (as->parms[7].items) {
	/* -cache */
	PrintCacheConfig(conn);
	return 0;
    }

    if (as->parms[8].items) {
	/* -cellservdb */
	PrintCellServDB(conn);
	return 0;
    }

    if (as->parms[5].items)
        print_ctime = 1;

    if (as->parms[2].items)
        /* -long */
	int32p = 1;
    else if (as->parms[3].items)
        /* -refcounts */
        int32p = 2;
    else if (as->parms[4].items)
        /* -callbacks */
        int32p = 4;
    else
	int32p = 0;

    if (int32p == 0 || int32p == 1)
        PrintLocks(conn, int32p);
    if (int32p >= 0 || int32p <= 4)
        PrintCacheEntries(conn, int32p);
    return 0;
}

#ifndef AFS_NT40_ENV
#include "AFS_component_version_number.c"
#endif

int
main(int argc, char **argv)
{
    register struct cmd_syndesc *ts;

#ifdef	AFS_AIX32_ENV
    /*
     * The following signal action for AIX is necessary so that in case of a 
     * crash (i.e. core is generated) we can include the user's data section 
     * in the core dump. Unfortunately, by default, only a partial core is
     * generated which, in many cases, isn't too useful.
     */
    struct sigaction nsa;

    sigemptyset(&nsa.sa_mask);
    nsa.sa_handler = SIG_DFL;
    nsa.sa_flags = SA_FULLDUMP;
    sigaction(SIGSEGV, &nsa, NULL);
#endif

#ifdef AFS_NT40_ENV
    if (afs_winsockInit() < 0) {
        printf("%s: Couldn't initialize winsock. Exiting...\n", argv[0]);
        return 1;
    }
#endif

    rx_Init(0);

    ts = cmd_CreateSyntax(NULL, CommandProc, NULL, "query afs cache manager");
    cmd_AddParm(ts, "-servers", CMD_SINGLE, CMD_REQUIRED, "server machine");
    cmd_AddParm(ts, "-port", CMD_SINGLE, CMD_OPTIONAL, "IP port");
    cmd_AddParm(ts, "-long", CMD_FLAG, CMD_OPTIONAL, "print all info");
    cmd_AddParm(ts, "-refcounts", CMD_FLAG, CMD_OPTIONAL, 
                 "print only cache entries with positive reference counts");
    cmd_AddParm(ts, "-callbacks", CMD_FLAG, CMD_OPTIONAL, 
                 "print only cache entries with callbacks");
    cmd_AddParm(ts, "-ctime", CMD_FLAG, CMD_OPTIONAL, 
                "print human readable expiration time");

    
    cmd_AddParm(ts, "-addrs", CMD_FLAG, CMD_OPTIONAL,
		"print only host interfaces");
    cmd_AddParm(ts, "-cache", CMD_FLAG, CMD_OPTIONAL,
		"print only cache configuration");
    cmd_AddParm(ts, "-cellservdb", CMD_FLAG, CMD_OPTIONAL, 
                "print only cellservdb info");

    cmd_Dispatch(argc, argv);
    exit(0);
}
